vm/src/pyobject.rs

```diff
@@ -732,8 +731,8 @@ impl PyContext {
     }
 
     pub fn set_attr(&self, obj: &PyObjectRef, attr_name: &str, value: PyObjectRef) {
-        if let Some(PyModule { ref scope, .. }) = obj.payload::<PyModule>() {
-            scope.locals.set_item(self, attr_name, value)
+        if let Some(PyModule { ref dict, .. }) = obj.payload::<PyModule>() {
```

Comment:
Why is the module dict different then the pyobject dict? Can't we use obj.dict all the time and not have a special case for module?

Reply:
That's my eventual plan.

At the moment it's blocked because PyObject stores a PyAttributes rather than a real dict. Once I figure out how to resolve that the module attributes can move out and this special case goes away.

Reply:
(Making Class the special case might be the easiest short term fix. It's the one with the most special logic, high performance requirements, and it's impossible to get a reference to the underlying dict.)

Reply:
Why cannot we leave the `dict` member to `PyAttributes`? Still I feel that using a full dict is a performance thing, since we need to use hashes from strings to index them.

Reply:
How do we get this code to work without real dictionaries? (It works on this branch.)

```
g = globals()
g['x'] = 2
exec('x += 2')
assert x == 4
assert g['x'] == x
```

I think this is a case of make it correct, then make it fast. There are optimisations we can apply later to avoid hash lookups in a many cases.

Reply:
My idea was by having a sort of dict proxy type, a lightweight dict type which only supports indexing by strings. The ObjectPayload could be PyAttributes or PyDict. The type could be dict_type in both cases. We could do automatic conversion from one to the other type when required. For example, when a PyDict is given, but we require it to be PyAttributes, we convert it to PyAttributes. When we want to insert an integer as key, we can upgrade again from PyAttributes to PyDict.

That being said, I'm happy to go with the approach make it work, then make it fast :)

Reply:
I've thought about that sort of optimisation, and we could definitely do something like that.

CPython does something similar for locals and class attributes, which is a source of weird edge cases and bugs.